### ISSUE TITLE
Problems testing terra[ Release 1.0.0-beta1 ]

### ISSUE THREAD

**Issue by tuxd3v** [terralang/terra#366]
I downloaded the Binaryes files..for x86_64:
terra-Linux-x86_64-2e2032e.zip

extracted them, then
cd share/terra/tests
../../../bin/terra ./run

lots of errors..
```lua
=================
= FAILING tests
=================
output.t
includec.t
linkllvm.t
luaapi.t
exportdynamic.t
dynlib.t
=================

517 tests passed. 6 tests failed.
```
the same hapens with:

```lua
ls -l /usr/local/{bin/terra,lib/terra/terra.so}
lrwxrwxrwx 1 root staff        17 Apr 13 18:32 /usr/local/bin/terra -> terra.1.0.0-beta1
-rwxr-xr-x 1 root staff 116470296 Apr 14 14:37 /usr/local/lib/terra/terra.so
```

**Comment by elliottslaughter** [comment]
We should probably relabel the Linux binaries as Ubuntu 16.04, because it seems they are not as portable as we'd like. If you're on another distro, my first recommendation would be to try building from source. Unfortunately, much (most?) of the incompatibility seems to come from Clang, so I'm not sure we're in a position to do anything about it.

If you have trouble building from source please report this along with your distro and compiler versions, etc.

**Comment by aiverson** [comment]
I've started writing some Dockerfiles to get portable executables and to build minimalistic containers for Terra. That should help.

**Comment by tuxd3v** [comment]
Thanks @elliottslaughter,
I will try when I have some spare time again , compiling from Sources :)

Regards,

**Comment by jane3ek** [comment]
I was trying to use Release 1.0.0-beta1 on Centos 7. The problem which I found was a reference to a specific version of libtinfo in terra library which is causing linking errors and stopping from building an executable using that release:

```
 ../../../bin/terra: /lib64/libtinfo.so.5: no version information available (required by ../../../bin/terra)
 terra-Linux-x86_64-2e2032e/lib
 -Wl,-rpath,terra-Linux-x86_64-2e2032e/lib terra-Linux-x86_64-2e2032e/lib/terra.so
 terra-Linux-x86_64-2e2032e/lib/terra.so: undefined reference to setupterm@NCURSES_TINFO_5.0.19991023
 terra-Linux-x86_64-2e2032e/lib/terra.so: undefined reference to del_curterm@NCURSES_TINFO_5.0.19991023
 terra-Linux-x86_64-2e2032e/lib/terra.so: undefined reference to tigetnum@NCURSES_TINFO_5.0.19991023
 terra-Linux-x86_64-2e2032e/lib/terra.so: undefined reference to set_curterm@NCURSES_TINFO_5.0.19991023
```

On my distro libtinfo is on 5.9 version and does not link (5.0 is not available). Is there an option to deliver a build without reference to specific library version? 
If possible I would like to avoid building it from sources (especially that previous release 'release-2016-03-25' works fine without those errors)

**Comment by elliottslaughter** [comment]
It may have been that the old binary was built with a custom LLVM that specifically excluded those dependencies. Or it may be that Ubuntu 14.04's version of those dependencies was more compatible with CentOS 7.

It would be a bit of a pain, but we could do a custom LLVM build to reduce the required dependencies. The main issue is that we're likely to see timeouts in Travis for our LLVM builds. But we at least know what the flags should be: https://github.com/zdevito/terra/issues/326

Anyway, I think it's still important to set realistic expectations about how portable the binaries can be. We could expand the table of distros we build for, but I think Clang makes no promises about being portable (at a binary level) and so we can't really make any promises either.

**Comment by aiverson** [comment]
Maybe we should consider using Holy Build Box. I've been using an alpine docker image to build and test terra for a bit. I think between an alpine docker image/build and a holy build box set of binaries we would get most of the useful portability.

**Comment by aiverson** [comment]
Does anyone have a problem with me making up docker scripts for building for alpine and portable normal linuxes? Does that seem useful?

**Comment by elliottslaughter** [comment]
I don't know what a Holy Build Box is, but I agree that Docker is probably the only sane way to support multiple distros. My one wish list item would be that it integrate with the CI so we can have it in both our regular test suite as well as for releases (though perhaps with a smaller matrix on non-release commits to keep CI times reasonable). We can potentially ship both Docker images and binaries, so people can pick whatever format they find useful.

**Comment by aiverson** [comment]
Holy build box is a docker container specifically designed to be used to compile software to be compatible with as wide a range of linuxes as possible. It has old versions of glibc, statically linked versions of a bunch of libraries, configuration settings to minimize the amount of assumptions that get grabbed by the build tools, etc. Basically, you just use docker to launch the holy build box image with the application's source directory mapped as a volume and a compilation script, then all the compilation is performed in a clean environment. [The github page](https://github.com/phusion/holy-build-box)

**Comment by OvermindDL1** [comment]
The Holy Build Box is awesome, I highly recommend it as well.

**Comment by elliottslaughter** [comment]
@tuxd3v Can you at least tell us what distro and version you were on when you tested this, so that we can check if any of our changes are helping? Thanks.

**Comment by elliottslaughter** [comment]
I finally got fed up enough by this to go about trying to quantify exactly how portable we are.

The initial results were horrific. With the `docker/Dockerfile.ubuntu` configuration (i.e., using LLVM from Ubuntu's package repositories), we got exactly zero cross-version compatibility. Ubuntu 18.04 binaries don't run on 20.04, and 20.04 binaries don't run on 22.04. Errors very reminiscent of https://github.com/terralang/terra/issues/366#issuecomment-493871533 above. Don't even think about cross-distro.

I fiddled with that a bit but ultimately concluded it was hopeless because the pre-packaged LLVM includes too many dependencies. There simply is no way to make it compatible in the presence of so many additional libraries. The story would basically be the same for any distro, even a minimalist distro, unless it provides a set of minimal-dependency LLVM binaries for us to use. The only way forward is to cut those dependencies out; and the only way I can see to cut them out is to build from source.

So I added the code to build from source. This is currently sitting in #537. It will probably be extremely slow. But the compatibility results are promising: Ubuntu 18.04 binaries successfully load and run all tests in 20.04 and 22.04. As a bonus, I also reduced the set of runtime dependencies that need to be installed on the target machine.

I don't know how the cross-distro compatibility will look. libc and libstdc++ may still be an issue there, not to mention embedded headers. But I think there's only so much we can expect from this, and anyway there's only so much effort that I'm personally able to put into getting this to work cross-distro.

**Comment by elliottslaughter** [comment]
I merged #537 and am going to consider this fixed. There is still #388 if we want to go that direction; but I think the minimal Ubuntu build is actually pretty good and ought to work on modern-ish distros (maybe not on older RHEL/CentOS, but I think that's ok). Certainly it seems future-proof if you stay within the Ubuntu family.
